tests/bsim/bluetooth/mesh/src/test_beacon.c

```diff
@@ -1338,38 +1357,393 @@ static void test_rx_priv_interleave(void)
+				    &same_random));
 	ASSERT_EQUAL(0x03, beacon.flags);
 	ASSERT_EQUAL(0x0001, beacon.iv_index);
 
 	PASS();
 }
+
+#if CONFIG_BT_MESH_GATT_PROXY
+
+#define BEACON_TYPE_PRIVATE_NET_ID  2
+#define BEACON_TYPE_PRIVATE_NODE_ID 3
+#define BEACON_TYPE_PRIVATE_LEN     28
+#define TEST_NET_IDX1               0
+#define TEST_NET_IDX2               1
+#define MAX_TIMEOUT ((CONFIG_BT_MESH_NODE_ID_TIMEOUT * 1000) / 6)
+
+struct pp_netkey_ctx {
+	uint8_t *net_key;
+	uint8_t net_id[8];
+	struct bt_mesh_key id_key;
+};
+
+static struct pp_netkey_ctx pp_net0 = {.net_key = (uint8_t *)test_net_key};
+static struct pp_netkey_ctx pp_net1 = {.net_key = (uint8_t *)test_net_key_secondary};
+
+struct priv_test_ctx {
+	uint8_t beacon_type;
+	uint16_t *node_id_addr;
+	union {
+		bool same_random;
+		uint8_t itr;
+	} arg;
+};
+
+static uint64_t last_pp_random;
+
+static void pp_netkey_ctx_init(struct pp_netkey_ctx *net)
+{
+	int err;
+
+	err = bt_mesh_identity_key(net->net_key, &net->id_key);
+	if (err) {
+		FAIL("Failed to generate ID key (err %d)", err);
+	}
+
+	err = bt_mesh_k3(net->net_key, net->net_id);
+	if (err) {
+		FAIL("Failed to generate Net ID (err %d)", err);
+	}
+}
+
+static bool pp_type_check(uint16_t expected_beacon, uint8_t adv_type, struct net_buf_simple *buf)
+{
+	if (adv_type != BT_GAP_ADV_TYPE_ADV_IND || buf->len != BEACON_TYPE_PRIVATE_LEN) {
+		return false;
+	}
+
+	/* Remove Header */
+	(void)net_buf_simple_pull_mem(buf, 11);
+
+	uint8_t beacon_type = net_buf_simple_pull_u8(buf);
+
+	if (beacon_type != expected_beacon) {
+		return false;
+	}
+
+	return true;
+}
+
+static uint64_t pp_hash_calc(struct pp_netkey_ctx *net, uint64_t random, uint16_t *addr)
+{
+	uint64_t hash;
+	uint8_t tmp[16] = {0, 0, 0, 0, 0, 3};
+
+	if (addr) {
+		memcpy(&tmp[6], &random, 8);
+		sys_put_be16(*addr, &tmp[14]);
+
+	} else {
+		memcpy(&tmp[0], net->net_id, 8);
+		memcpy(&tmp[8], &random, 8);
+	}
+
+	bt_mesh_encrypt(&net->id_key, tmp, tmp);
+	memcpy(&hash, &tmp[8], 8);
+
+	return hash;
+}
+
+static bool pp_beacon_check(const uint8_t *net_id, void *ctx)
+{
+	bool ret;
+	struct priv_test_ctx *test_ctx = (struct priv_test_ctx *)ctx;
+
+	ASSERT_EQUAL(beacon.pp_hash,
+		     pp_hash_calc(&pp_net0, beacon.pp_random, test_ctx->node_id_addr));
+
+	if (memcmp(beacon.adv_addr.a.val, last_beacon_adv_addr.a.val, BT_ADDR_SIZE) == 0) {
+		return false;
+	}
+
+	memcpy(&last_beacon_adv_addr.a.val, beacon.adv_addr.a.val, BT_ADDR_SIZE);
+
+	ret = test_ctx->arg.same_random ? (last_pp_random == beacon.pp_random)
+			  : (last_pp_random != beacon.pp_random);
+	last_pp_random = beacon.pp_random;
+
+	return ret;
+}
+
+
+static bool pp_multi_net_check(const uint8_t *net_id, void *ctx)
+{
+	struct priv_test_ctx *test_ctx = (struct priv_test_ctx *)ctx;
+	static uint8_t old_idx = 0xff;
+	static struct {
+		struct pp_netkey_ctx *net;
+		uint16_t recv_cnt;
+		int64_t start;
+	} net_ctx[2] = {
+		{.net = &pp_net0},
+
+		{.net = &pp_net1},
+	};
+
+	for (size_t i = 0; i < ARRAY_SIZE(net_ctx); i++) {
```

Comment:
Could this logic be inside `test_rx_priv_multi_net_id`? Like here but with 1 second timeout: https://github.com/zephyrproject-rtos/zephyr/blob/0004680d8a6b324d98ed1bca05b319a6668b79c6/tests/bsim/bluetooth/mesh/src/test_beacon.c#L954-L963